File: pippip-native-ios/Messages/PreviewCell.m
//
//  PreviewCell.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 2/3/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "pippip_native_ios-Swift.h"
#import "PreviewCell.h"
#import "Notifications.h"

static const NSInteger SEC_PER_HOUR = 3600;
static const NSInteger ONE_DAY = SEC_PER_HOUR * 24;

@interface PreviewCell ()
{
    TextMessage *textMessage;
    ContactManager *contactManager;
    BOOL configured;
}

@property (weak, nonatomic) IBOutlet UIImageView *messageReadImage;
@property (weak, nonatomic) IBOutlet UILabel *senderLabel;
@property (weak, nonatomic) IBOutlet UILabel *dateTimeLabel;
@property (weak, nonatomic) IBOutlet UILabel *previewLabel;

@end

@implementation PreviewCell

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    contactManager = [[ContactManager alloc] init];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(cleartextAvailable:)
                                                 name:CLEARTEXT_AVAILABLE
                                               object:nil];

}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];
    
    // Configure the view for the selected state
    configured = NO;

}

- (void)dealloc {

    [[NSNotificationCenter defaultCenter] removeObserver:self name:CLEARTEXT_AVAILABLE object:nil];

}

- (void)configure:(TextMessage*)message {

    [_messageReadImage setHidden:message.read];
    if (!configured) {
        configured = YES;
        textMessage = message;
        Contact *contact = [contactManager getContactById:message.contactId];
        _senderLabel.text = contact.displayName;
        NSString *dt = [self convertTimestamp:message.timestamp];
        _dateTimeLabel.text = [dt stringByAppendingString:@" >"];
        dispatch_async(dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0), ^{
            [message decrypt:false];    // noNotify = false
        });
    }

}

- (NSString*)convertTimestamp:(NSInteger)timestamp {

    NSInteger ts = timestamp / 1000;
    NSInteger now = [[NSDate date] timeIntervalSince1970];
    NSInteger elapsed = now - ts;
    NSInteger elapsedSinceMidnight = now % ONE_DAY;

    // Yay C!
    struct tm *ts_tm = localtime(&ts);
    char buf[50];
    if (elapsed <= elapsedSinceMidnight) {
        strftime(buf, 50, "%H:%S", ts_tm);
        return [NSString stringWithUTF8String:buf];
    }
    if (elapsed < ONE_DAY) {
        return @"Yesterday";
    }
    else {
        strftime(buf, 50, "%D", ts_tm);   // Localize this for Europe
        return [NSString stringWithUTF8String:buf];
    }

}

- (TextMessage*)getTextMessage {
    return textMessage;
}

- (void)setMessageText:(NSString*)msgText {

    if (msgText.length > 33) {
        NSString *preview = [msgText substringWithRange:NSMakeRange(0, 33)];
        _previewLabel.text = [preview stringByAppendingString:@"..."];
    }
    else {
        _previewLabel.text = msgText;
    }

}

- (void)cleartextAvailable:(NSNotification*)notification {

    TextMessage *message = (TextMessage*)notification.object;
    if (message.messageId == textMessage.messageId) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self setMessageText:self->textMessage.cleartext];
        });
    }

}
     
@end

File: pippip-native-ios/Messages/PreviewCell.swift
//
//  PreviewCell.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 5/2/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import ChameleonFramework

class PreviewCell: PippipTableViewCell {

    static let secondsPerHour: Int64 = 3600
    static let secondsPerDay: Double = 3600 * 24

    @IBOutlet weak var messageReadIndicator: UIImageView!
    @IBOutlet weak var senderLabel: UILabel!
    @IBOutlet weak var previewLabel: UILabel!
    @IBOutlet weak var timestampLabel: UILabel!

    var contactManager = ContactManager.instance
    var textMessage: TextMessage?

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
//        NotificationCenter.default.addObserver(self, selector: #selector(sessionEnded(_:)),
//                                               name: Notifications.SessionEnded, object: nil)

    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    func configure(textMessage: TextMessage) {

        messageReadIndicator.isHidden = textMessage.read
        if self.textMessage == nil || self.textMessage?.messageId != textMessage.messageId {
            NotificationCenter.default.addObserver(self, selector: #selector(cleartextAvailable(_:)),
                                                   name: Notifications.CleartextAvailable, object: nil)
            self.textMessage = textMessage
            let contact = contactManager.getContact(contactId: textMessage.contactId)
            senderLabel.text = contact?.displayName
            timestampLabel.text = convertTimestamp(textMessage.timestamp) + " >"
            if textMessage.ciphertext!.count < 25 {
                textMessage.decrypt(notify: false)   // No notification
                setPreviewText(textMessage.cleartext ?? "Text not available")
            }
            else {
                DispatchQueue.global(qos: .background).async {
                    textMessage.decrypt(notify: true)
                }
            }
        }

        //senderLabel.textColor = PippipTheme.darkTextColor
        //previewLabel.textColor = PippipTheme.darkTextColor
        //timestampLabel.textColor = PippipTheme.darkTextColor

        super.setMediumTheme()

    }

    func convertTimestamp(_ timestamp: Int64) -> String {

        let messageDate = Date(timeIntervalSince1970: Double(timestamp / 1000))
        let now = Date()
        let elapsed = now.timeIntervalSince(messageDate)

        if elapsed > PreviewCell.secondsPerDay {
            let formatter = DateFormatter()
            formatter.dateStyle = .short
            formatter.timeStyle = .none
            return formatter.string(from: messageDate)
        }
        else {
            let formatter = DateFormatter()
            formatter.dateStyle = .short
            formatter.timeStyle = .short
            return formatter.string(from: messageDate)
        }

    }

    func setPreviewText(_ text: String?) {

        guard let _ = text else { return }
        if text!.utf8.count > 33 {
            previewLabel.text = text!.prefix(33) + " ..."
        }
        else {
            previewLabel.text = text
        }

    }

    @objc func cleartextAvailable(_ notification: Notification) {

        if let textMessage = notification.object as? TextMessage {
            if textMessage.messageId == self.textMessage?.messageId {
                NotificationCenter.default.removeObserver(self, name: Notifications.CleartextAvailable, object: nil)
                DispatchQueue.main.async {
                    let text = textMessage.cleartext ?? "Text Not Available"
                    self.setPreviewText(text)
                }
            }
        }

    }

//    @objc func sessionEnded(_ notification: Notification) {
//
//        textMessage = nil
//
//    }

}
